This tutorial is based on the [Library Carpentry SQL tutorial](https://librarycarpentry.github.io/lc-sql/), licensed under [CC-BY 4.0](https://librarycarpentry.github.io/lc-sql/LICENSE.html).

You can run this Jupyter notebook in [Google Colab](https://colab.research.google.com/) or on your local installation.

**Note**: In Colab, if the notebook is View Only, remember to make a copy, so that you can modify it:

* _File --> Save a copy in Drive_.


## Attribution

Jordan Perdersen (Ed.), Kristin Lee (Ed.), Christopher Erdmann (Ed.), Lise Doucette (Ed.), Elaine Wong (Ed.), Janice Chan (Ed.), James Baker, Fernando Rios, Tim Dennis, Belinda Weaver, … orobecca. (2019, July). **LibraryCarpentry/lc-sql: Library Carpentry: Introduction to SQL, June 2019** (Version v2019.06.1). _Zenodo_. http://doi.org/10.5281/zenodo.3266102

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3266102.svg)](https://doi.org/10.5281/zenodo.3266102)

### Modifications

Extracts modified to Jupyter notebook use by Stian Soiland-Reyes <https://orcid.org/0000-0001-9842-9718>.

### References

* <https://pythonspot.com/sqlite-database-with-pandas/>
* <https://pythoncourses.gumroad.com/l/KmxqY>
* <https://www.sqlitetutorial.net/>
* <https://wesmckinney.com/book/pandas-basics>

# Data ingestion

In this tutorial we assume data has already been ingested and cleaned. We'll retrieve the already prepared SQLite database.

[SQLite](https://sqlite.org/) is a miniature relational database that can be used without a server installation, for instance by directly importing the SQLite library in Python and "connecting" it to a local file. It is worth noting that this means only a single process can access the file at a time (in this case our Jupyter notebook).

## Download SQLite database

The Zenodo deposit <https://doi.org/10.5281/zenodo.8360812> contains a ZIP file with the pre-populated database in SQLite format. As this tutorial is not showing "proper" data ingestion, we'll just download and unzip using Linux shell commands:



In [ ]:
!wget https://zenodo.org/records/8360812/files/doaj-article-sample-db.zip && unzip doaj-article-sample-db.zip

--2025-07-08 15:12:03--  https://zenodo.org/records/8360812/files/doaj-article-sample-db.zip
Resolving zenodo.org (zenodo.org)... 188.185.45.92, 188.185.48.194, 188.185.43.25, ...
Connecting to zenodo.org (zenodo.org)|188.185.45.92|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 947007 (925K) [application/octet-stream]
Saving to: ‘doaj-article-sample-db.zip’

doaj-article-sample 100%[===================>] 924.81K   499KB/s    in 1.9s    

2025-07-08 15:12:06 (499 KB/s) - ‘doaj-article-sample-db.zip’ saved [947007/947007]

Archive:  doaj-article-sample-db.zip
  inflating: doaj-article-sample.db  
  inflating: doaj-article-sample.db.sql  


Now let's connect to this database from within Python using the [sqlite3](https://docs.python.org/3/library/sqlite3.html) library, and see which tables are included. In SQLite this can be done by inspecting the `sqlite_master` table, using the `SELECT *` query . We'll use [pandas](https://pandas.pydata.org/) as a helper to better view the SQL results in Jupyter Notebook.



In [ ]:
import sqlite3
import pandas as pd

db = sqlite3.connect("doaj-article-sample.db")
pd.read_sql_query("SELECT * FROM sqlite_master", db)

type             name         tbl_name  rootpage  \
0  table  sqlite_sequence  sqlite_sequence       275   
1  table        languages        languages       408   
2  table       publishers       publishers       679   
3  table         licences         licences         2   
4  table         journals         journals         3   
5  table         articles         articles         5   

                                                 sql  
0             CREATE TABLE sqlite_sequence(name,seq)  
1  CREATE TABLE "languages" (\n  "id" INTEGER NOT...  
2  CREATE TABLE "publishers" (\n  "id" INTEGER NO...  
3  CREATE TABLE "licences" (\n  "id" INTEGER NOT ...  
4  CREATE TABLE "journals" (\n  "id" INTEGER NOT ...  
5  CREATE TABLE "articles" (\n  "id" INTEGER NOT ...

# Introduction to SQL

In this tutorial we'll follow the Library Carpentry guide, but run the queries from within Python instead of the DB Browser.  

* Read through [Section 1: Introduction to SQL](https://librarycarpentry.github.io/lc-sql/01-introduction.html)  (ignoring the _DB Browser_ subsection).



# Selecting and querying data

Next, follow the [Section 2: Selecting and sorting data](https://librarycarpentry.github.io/lc-sql/02-selecting-sorting-data.html). Use Python's `"""` strings to support multiple lines.

Note that you don't need to use `print()`, as Jupyter Notebook will output the result of the last line, which with Panda becomes a HTML table:

In [ ]:
pd.read_sql_query("""
SELECT title
FROM articles
""", db)

Title
0      The Fisher Thermodynamics of Quasi-Probabilities
1     Aflatoxin Contamination of the Milk Supply: A ...
2     Metagenomic Analysis of Upwelling-Affected Bra...
3     Synthesis and Reactivity of a Cerium(III) Scor...
4     Performance and Uncertainty Evaluation of Snow...
...                                                 ...
996   Crystal structure of [3-(1H-benzimidazol-2-yl)...
997   Crystal structure of bis(3-bromopyridine-κN)bi...
998   Crystal structure of 4,4′-(ethane-1,2-diyl)bis...
999   Crystal structure of (Z)-4-[1-(4-acetylanilino...
1000  Metagenomic Analysis of Upwelling-Affected Bra...

[1001 rows x 1 columns]

## SQL results in Python

Above, Pandas' DataFrame is rendered in Juyter Notebook, skipping most of the 1001 rows. Alternatively we can inspect the table returned in Python. This is beneficial in case you need to do further processing in code.

**Note**: You do _not_ have to use Panda dataframes for accessing SQL databases from Python, see also the [SQLite Python tutorial](https://docs.python.org/3/library/sqlite3.html#tutorial) for accessing the database directly.

Below, the method `.head()` is used to only select the first couple lines of the dataframe, while `.itertuples()` iterates over each row in the SQL result.  See the [Getting started with pandas](https://wesmckinney.com/book/pandas-basics) for further guidance on dataframes.

**Warning**: While SQL is generally case-insensitive, the database's spelling of the column name will be returned, which in this case is `Title` (instead of `title` as in the query) -- this matters below as Python code is case-sensitive.


In [ ]:
articles = pd.read_sql_query("""
SELECT Title
FROM articles
""", db)
for article in articles.head().itertuples():
  print(article.Title)


The Fisher Thermodynamics of Quasi-Probabilities
Aflatoxin Contamination of the Milk Supply: A Pakistan Perspective
Metagenomic Analysis of Upwelling-Affected Brazilian Coastal Seawater Reveals Sequence Domains of Type I PKS and Modular NRPS
Synthesis and Reactivity of a Cerium(III) Scorpionate Complex Containing a Redox Non-Innocent 2,2′-Bipyridine Ligand
Performance and Uncertainty Evaluation of Snow Models on Snowmelt Flow Simulations over a Nordic Catchment (Mistassibi, Canada)



**Tip**: You can modify the code of Jupyter notebook, and click each cell's ▶️ button to run them. Use the [+ Code] button in Colab to add further code blocks. Use one code block per output.




## Errors

Before we continue, let's look quickly at different errors in the SQL, which will cause a Python exception.

### Syntax error

The Python code is valid, but the SQL inside has a typo (`SELET` instead of `SELECT`):

In [ ]:
pd.read_sql_query("""
SELET title
FROM articles
""", db)

DatabaseError: Execution failed on sql '
SELET title
FROM articles
': near "SELET": syntax error

### Semantic error

Here the query is correct syntactically, but it's referencing a column that does not exist:

In [ ]:
pd.read_sql_query("""
SELECT wrongColumn
FROM articles
""", db)

DatabaseError: Execution failed on sql '
SELECT wrongColumn
FROM articles
': no such column: wrongColumn

* First, try to correct the errors above, then re-execute the cells.

Next, follow the [rest of section 2](https://librarycarpentry.github.io/lc-sql/02-selecting-sorting-data.html#capitalization-and-good-style) onwards to the remaining SQL selection exercises.


## Capitalization and good style

_Follow exercise on [good style](https://librarycarpentry.github.io/lc-sql/02-selecting-sorting-data.html#capitalization-and-good-style) below:_

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

## Unique values

_Follow exercise on [unique values](https://librarycarpentry.github.io/lc-sql/02-selecting-sorting-data.html#unique-values)_

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

## Sorting

_Follow exercise on [sorting](https://librarycarpentry.github.io/lc-sql/02-selecting-sorting-data.html#sorting)_


In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

# Filtering

_Follow exercies on [filtering](https://librarycarpentry.github.io/lc-sql/03-filtering.html)_

In [ ]:
pd.read_sql_query("""
SELECT *
FROM articles
WHERE ISSNs = '2056-9890'
""", db).head()

id                                              Title  \
0  142  Crystal structure of 7-isopropyl-1,4a,N-trimet...   
1  143  Crystal structure of 2-(2,4-diphenyl-3-azabicy...   
2  144  Crystal structure of 2-[2-(hydroxyimino)-1-phe...   
3  145  Crystal structure of 1-methoxy-5-methyl-N-phen...   
4  146  Redetermined structure of 4,4′-bipyridine–1,4-...   

                                             Authors  \
0                   Li Liu|Xin-Yan Yan|Xiao-Ping Rao   
1   K. Priya|K. Saravanan|S. Kabilan|S. Selvanayagam   
2  Brian J. Anderson|Michael B. Freedman|Sean P. ...   
3  Inna S. Khazhieva|Tatiana V. Glukhareva|Pavel ...   
4                       Rima Paul|Sanchay Jyoti Bora   

                         DOI  \
0  10.1107/S2056989015017648   
1  10.1107/S2056989015017740   
2  10.1107/S2056989015017739   
3  10.1107/S2056989015017776   
4  10.1107/S2056989015017569   

                                                 URL  \
0  https://doaj.org/article/69bfe1351c864bc8a6b38...   
1  https://doaj.org/article/ce085e7ac3bd43e49a122...   
2  https://doaj.org/article/4dd1a69aaefa478b8d2bb...   
3  https://doaj.org/article/8c3bd9d71ba642f5b28d0...   
4  https://doaj.org/article/b2d1ca2bb6fb4ba48c11e...   

                                            Subjects      ISSNs  \
0  crystal structure|dihydroabietic acid derivati...  2056-9890   
1  crystal structure|3-azabicyclononane derivativ...  2056-9890   
2  crystal structure|thiosemicarbazone|weak inter...  2056-9890   
3  crystal structure|1,2,3-triazole|rearrangement...  2056-9890   
4  crystal structure|co-crystal|supramolecular in...  2056-9890   

                                            Citation  LanguageId  LicenceId  \
0  Acta Crystallographica Section E: Crystallogra...           1          1   
1  Acta Crystallographica Section E: Crystallogra...           1          1   
2  Acta Crystallographica Section E: Crystallogra...           1          1   
3  Acta Crystallographica Section E: Crystallogra...           1          1   
4  Acta Crystallographica Section E: Crystallogra...           1          1   

   Author_Count       First_Author  Citation_Count  Day  Month  Year  
0             3             Li Liu              10    1     10  2015  
1             4           K. Priya              10    1     10  2015  
2             5  Brian J. Anderson              10    1     10  2015  
3             4  Inna S. Khazhieva              10    1     10  2015  
4             2          Rima Paul              10    1     10  2015

Note that in SQL, value comparisons is done with a _single_ equal sign  `=`, unlike Python and other programing languagues typical use of `==` or `.equals`

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db).head()

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db).head()

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db).head()

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db).head()

# Ordering and commenting

_Follow exercises on [Ordering and commenting](https://librarycarpentry.github.io/lc-sql/04-ordering-commenting.html)_





In [ ]:
pd.read_sql_query("""
SELECT Title, Authors
FROM articles
WHERE ISSNs = '2067-2764|2247-6202'
ORDER BY First_Author ASC;
""", db)

Title  \
0   Zweier I-Convergent Double Sequence Spaces Def...   
1   New Čebyšev Type Inequalities for Functions wh...   
2   Initial Maclaurin Coecients Bounds for New Sub...   
3   Measure of Tessellation Quality of Voronoï Meshes   
4   The Applicability of $-Calculus to Solve Some ...   
5   Some Families of q-Series Identities and Assoc...   
6   Lp - Approximation of Analytic Functions on Co...   
7   A Mixed Integer Linear Programming Formulation...   
8              Properties of Stabilizing Computations   
9   Primality Testing and Factorization by using F...   
10                                 Luhn Prime Numbers   
11  Hadamard Product of Certain Harmonic Univalent...   
12  Second Hankel Determinant for Generalized Saka...   
13  Some Perturbed Ostrowski Type Inequalities for...   
14  Katsaras’s Type Fuzzy Norm under Triangular Norms   
15  On BVσ I-convergent Sequence Spaces Defined by...   
16  Coupled Systems of Fractional Integro-Differen...   

                                              Authors  
0                A. Khan Vakeel| Khan Nazneen|Yasmeen  
1                            B. Meftah|K. Boukerrioua  
2                   Basem Aref Frasin|Tariq Al-Hawary  
3                               E. A-iyeh|J.F. Peters  
4                                     Eugene Eberbach  
5            H. M. Srivastava|S. N. Singh|S. P. Singh  
6                          Kumar Devendra|Jain Vandna  
7                                     Marija Ivanović  
8                                         Mark Burgin  
9                                       Musha Takaaki  
10                Octavian Cira|Florentin Smarandache  
11                       R. M. El-Ashwah|B. A. Frasin  
12                S. P. Vijayalakshmi|T. V. Sudharsan  
13                                     S. S. Dragomir  
14  Sorin Nădăban|Tudor Bînzar|Flavius Pater|Carme...  
15  Vakeel A. Khan|Mohd Shafiq|Rami Kamel Ahmad Ra...  
16  Zoubir Dahmani|Mohamed Amin Abdellaoui|Mohamed...

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

# Aggregating values

_Follow exercise on [Aggregating and calculating values](https://librarycarpentry.github.io/lc-sql/05-aggregating-calculating.html)_




In [ ]:
pd.read_sql_query("""
SELECT ISSNs, AVG(Citation_Count)
FROM articles
GROUP BY ISSNs;
""", db)

ISSNs  AVG(Citation_Count)
0   0367-0449|1988-3250             4.000000
1             1099-4300             4.000000
2             1420-3049             4.000000
3             1422-0067             8.000000
4             1424-8220             4.000000
5             1660-3397             4.000000
6             1660-4601            10.000000
7             1996-1944             4.000000
8             1999-4907             4.000000
9             1999-4915             4.000000
10            2056-9890             9.593240
11  2067-2764|2247-6202             8.647059
12            2071-1050             5.000000
13            2072-4292             5.000000
14            2072-6643             4.000000
15            2072-6651             4.000000
16            2072-6694             4.000000
17            2073-4344             4.000000
18            2073-4360             4.000000
19            2073-4395             4.000000
20            2073-4425             4.000000
21            2073-4441             4.000000
22            2073-8994             4.000000
23            2075-163X             4.000000
24            2075-1702             4.000000
25            2075-4418             4.000000
26            2075-4442             4.000000
27            2075-4701             4.000000
28            2075-5309             4.000000
29            2076-0787             4.000000
30            2076-2615             4.000000
31            2076-3417             5.000000
32            2076-3905             4.000000
33            2076-393X             4.000000
34            2077-0375             4.000000
35            2077-0472             5.000000
36            2077-1444             4.000000
37            2078-1547             4.000000
38            2079-6374             4.000000
39            2079-7737             4.000000
40            2079-8954             4.000000
41            2079-9268             8.000000
42            2220-9964             8.000000
43            2227-9717             4.000000
44            2250-1177             7.000000
45  2278-4748|2278-4802             8.000000
46            2304-6740             5.000000
47            2304-6767             5.000000
48            2305-6304             3.000000
49            2306-5338             4.000000
50            2306-5354             5.000000

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db)

# Joins and aliases

_Follow exercises on [Joins and aliases](https://librarycarpentry.github.io/lc-sql/06-joins-aliases.html)_



In [ ]:
pd.read_sql_query("""
SELECT *
FROM articles
JOIN journals
ON articles.ISSNs = journals.ISSNs;
""", db).head()


id                                              Title  \
0   0   The Fisher Thermodynamics of Quasi-Probabilities   
1   1  Aflatoxin Contamination of the Milk Supply: A ...   
2   2  Metagenomic Analysis of Upwelling-Affected Bra...   
3   3  Synthesis and Reactivity of a Cerium(III) Scor...   
4   4  Performance and Uncertainty Evaluation of Snow...   

                                             Authors  \
0                     Flavia Pennini|Angelo Plastino   
1                         Naveed Aslam|Peter C. Wynn   
2  Rafael R. C. Cuadrat|Juliano C. Cury|Alberto M...   
3  Fabrizio Ortu|Hao Zhu|Marie-Emmanuelle Boulon|...   
4  Magali Troin|Richard Arsenault|François Brissette   

                          DOI  \
0           10.3390/e17127853   
1  10.3390/agriculture5041172   
2       10.3390/ijms161226101   
3   10.3390/inorganics3040534   
4    10.3390/hydrology2040289   

                                                 URL  \
0  https://doaj.org/article/b75e8d5cca3f46cbbd63e...   
1  https://doaj.org/article/0edc5af6672641c0bd456...   
2  https://doaj.org/article/d9fe469f75a0442382b84...   
3  https://doaj.org/article/95606ed39deb4f43b96f7...   
4  https://doaj.org/article/18b1d70730d44573ab5c2...   

                                            Subjects      ISSNs  \
0  Fisher information|quasi-probabilities|complem...  1099-4300   
1  aflatoxins|AFM1|AFB1|milk marketing chains|hep...  2077-0472   
2  PKS|NRPS|metagenomics|environmental genomics|u...  1422-0067   
3  lanthanide|cerium|scorpionate|tris(pyrazolyl)b...  2304-6740   
4  snow models|hydrological models|snowmelt|uncer...  2306-5338   

                                            Citation  LanguageId  LicenceId  \
0       Entropy, Vol 17, Iss 12, Pp 7848-7858 (2015)           1          1   
1  Agriculture (Basel), Vol 5, Iss 4, Pp 1172-118...           1          1   
2  International Journal of Molecular Sciences, V...           1          1   
3  Inorganics (Basel), Vol 3, Iss 4, Pp 534-553 (...           1          1   
4         Hydrology, Vol 2, Iss 4, Pp 289-317 (2015)           1          1   

   ...          First_Author Citation_Count  Day  Month  Year  id     ISSN-L  \
0  ...        Flavia Pennini              4    1     11  2015  14  1099-4300   
1  ...          Naveed Aslam              5    1     11  2015   1  2077-0472   
2  ...  Rafael R. C. Cuadrat              8    1     11  2015  22  1422-0067   
3  ...         Fabrizio Ortu              5    1     11  2015  20  2304-6740   
4  ...          Magali Troin              4    1     11  2015  19  2306-5338   

       ISSNs PublisherId                                Journal_Title  
0  1099-4300           2                                      Entropy  
1  2077-0472           2                                  Agriculture  
2  1422-0067           2  International Journal of Molecular Sciences  
3  2304-6740           2                                   Inorganics  
4  2306-5338           2                                    Hydrology  

[5 rows x 21 columns]

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db).head()


In [ ]:
pd.read_sql_query("""
SELECT ...
""", db).head()


In [ ]:
pd.read_sql_query("""
SELECT ...
""", db).head()


In [ ]:
pd.read_sql_query("""
SELECT ...
""", db).head()


## Aliases

_Follow exercises on [aliases](https://librarycarpentry.github.io/lc-sql/06-joins-aliases.html#aliases)_

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db).head()


# Saving queries

_Follow exercise of [Saving queries](https://librarycarpentry.github.io/lc-sql/07-saving-queries.html)_

**Note**: As `DROP VIEW` and `CREATE VIEW` do not return any result, below we have to use `db.execute` directly rather than `pd.read_sql_query`.

In [ ]:
db.execute("DROP VIEW IF EXISTS journal_counts")

db.execute("""
CREATE VIEW ...

""")

In [ ]:
pd.read_sql_query("""
SELECT ...
""", db).head()

Remember that a view is not stored as a separate table, but is a pre-canned query that can then itself be queries. That means updates to the tables will modify the views! SQL Views typically cannot themselves be modified with `INSERT`.

Nevertheless, if a view was temporary, we may want to delete it so it does clutter our database:

In [ ]:
db.execute("DROP VIEW journal_counts")

# Additional reading

The remaining sections of Library Carpentry SQL relates to creating databases before they can be queried.


* [Database design](https://librarycarpentry.github.io/lc-sql/08-database-design.html)
* [Creating tables](https://librarycarpentry.github.io/lc-sql/09-create.html)
* [Extra challenges](https://librarycarpentry.github.io/lc-sql/11-extra-challenges.html)
* [Good SQL style](https://librarycarpentry.github.io/lc-sql/Bonus_GoodStyle.html)

**Tip**: Sometimes knowing what kind of queries are needed will influence the design at earlier data engineering phases, e.g. reducing the need for complex `JOIN`s.